### ISSUE TITLE
Option to generate endpoint synchronously on PUT /endpoint

### ISSUE THREAD

**Issue by tgraf** [cilium/cilium#4409]
Add a mode to synchronously generate the endpoint on the PUT /endpoint API call to allow ensuring that on CNI ADD, the endpoint is generated and ready by the time CNI ADD finished.

- [x] Add flag to API to enable sync mode.
- [x] Run with sync mode through CI to test feasibility of doing this by default

**Comment by aanm** [comment]
@tgraf FYI, the environment variable in the ConfigMap will not be used in the CNI binary invocation from kubelet.

**Comment by ianvernon** [comment]
The problem makes sense to me; we want to make sure that the networking is plumbed before CNI ADD returns. I looked at the CNI spec, and noticed that it doesn't specify anywhere that calls to add in CNI require that the networking plumbing are done when the API call finishes:

https://github.com/containernetworking/cni/blob/master/SPEC.md 

Specifically:

```
ADD: Add container to network

Parameters:
    * Container ID. A unique plaintext identifier for a container, allocated by the runtime. Must not be empty.
    * Network namespace path. This represents the path to the network namespace to be added, i.e. /proc/[pid]/ns/net or a bind-mount/link to it.
    * Network configuration. This is a JSON document describing a network to which a container can be joined. The schema is described below.
    * Extra arguments. This provides an alternative mechanism to allow simple configuration of CNI plugins on a per-container basis.
    * Name of the interface inside the container. This is the name that should be assigned to the interface created inside the container (network namespace); consequently it must comply with the standard Linux restrictions on interface names.
Result:
    * Interfaces list. Depending on the plugin, this can include the sandbox (eg, container or hypervisor) interface name and/or the host interface name, the hardware addresses of each interface, and details about the sandbox (if any) the interface is in.
    * IP configuration assigned to each interface. The IPv4 and/or IPv6 addresses, gateways, and routes assigned to sandbox and/or host interfaces.
    * DNS information. Dictionary that includes DNS information for nameservers, domain, search domains and options.
```

I am not saying that what is being asked in this ticket is wrong, but I am moreso wondering exactly what the requirements are of CNI's spec with respect to the plumbing of networking for the container by a plugin? Can you shed some light on this?

**Comment by tgraf** [comment]
I think the CNI specification is vague on this aspect on purpose as not all plugins may be able to provide connectivity synchronously. The purpose of this work is to better support workloads which assume network connectivity at container start, fail quickly within seconds and never retry.

**Comment by ianvernon** [comment]
Agreed. I do wonder how this will affect systems that are under temporary high load. Say for instance, that if the endpoint regeneration takes a long time (due to say, policy import or identity allocation right before a call to CNI ADD is made), then all endpoints will be queued for regeneration. That means that it may take longer for the endpoint for which the PUT request was made to regenerate, especially since regeneration holds locks for structures shared amongst all endpoints, and because generation of endpoints takes time (usually seconds to compile BPF). This could result in CNI ADD calls failing in cases where it did not before because we synchronously wait, vs. eventually achieving desired state. I guess this is there the `ConfigMap` option comes into play? 

TL;DR: I'm worried about now we are reliant upon something returning synchronously when that operation completing in the specified amount of time is entirely reliant upon state in the system which is highly probable (lots of endpoints + policy regenerations occurring).

**Comment by tgraf** [comment]
Kubelet will retry up to 10 or 20 times I think and even reschedule to
another node which would be positive behavior.

**Comment by ianvernon** [comment]
@tgraf given that we are merging with CNI providing this option by default, can I remove the two environment variable / ConfigMap items from the requirements list for this?

**Comment by tgraf** [comment]
> @tgraf given that we are merging with CNI providing this option by default, can I remove the two environment variable / ConfigMap items from the requirements list for this?

Yes, agreed.
